carp.common/src/commonMain/kotlin/dk/cachet/carp/common/MACAddress.kt

```diff
@@ -0,0 +1,49 @@
+package dk.cachet.carp.common
+
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+
+
+/**
+ * Represents a IEEE 802 48-bit media access control address (MAC address):
+ * a unique identifier assigned to a network interface controller (NIC) for use as a network address in communications within a network segment.
+ *
+ * This is equivalent to the EUI-48 identifier.
+ *
+ * @param address Six groups of two upper case hexadecimal digits, separated by hyphens (-).
+ *
+ * TODO: Allow more flexible [address] input and conversion to the common form. Make sure equality is ensured!
+ */
+@Serializable( with = MACAddressSerializer::class )
+data class MACAddress( val address: String )
+{
+    init
+    {
+        require( MACAddressRegex.matches( address ) )
+            { "Invalid MAC address string representation. Expected six groups of two upper case hexadecimal digits, separated by hyphens (-)." }
+    }
+}
+
+
+/**
+ * Regular expression to verify whether the string representation of a [MACAddress] matches the IEEE 802 standard notation.
+ */
+val MACAddressRegex = Regex( "([0-9A-F]{2}-){5}([0-9A-F]{2})" )
+
```

Comment:
May I suggest: `^([0-9A-Fa-f]{2}-){5}([0-9A-Fa-f]{2})$` IEEE actually states that `:` is also a valid delimiter but I'm fine with hyphen only. It should however be case insensitive

Reply:
What the 'official' format is was a bit ambiguous. I did find somewhere '-' was the main one recommended by IEEE. Upper case and lower case wasn't specified, but all examples were upper case.

Regardless, of course we want to support easier initialization, but the main point here is the TODO note at the top:

> TODO: Allow more flexible [address] input and conversion to the common form. Make sure equality is ensured!

Reply:
Fine with leaving the more strict format as is for now, but unless you add the meta characters matching the string boundaries _or_ require a string of the correct length, then the regex will match any string _containing_ a valid mac

Reply:
I did it in the meantime awaiting your answer. =)

I thought that as well (and a similar problem would exist for UUID validation), but that's why I added the 'too long' test:
`assertFailsWith<IllegalArgumentException> { MACAddress( "00-00-00-00-00-00-FF" ) }`

I think this specific API does full matching, but I'd have to verify.

Reply:
It passes because of incorrect group size, no. But groups are determined by the delimiter.

Reply:
If you just use the boundary matching regex you can leave the first require statement entirely

Reply:
Even better. You can just use my regex in its entirety. Your formatting will always ensure the field has the stricter format, but allow more relaxed input. win-win

Reply:
No. I also thought of that, but `... AB-CD-EF-12-34-56 ...` passes and are 6 groups. It passes because the `matches` API does not do what you think it does AFAIK.

The first require statement is there for different reasons; to disallow matching `AB:CD-EF:12-34-56`.

And, the regex is public. As is, the regex can be reused to also find matches _anywhere_ in a sentence. Adding ^ or $ wouldn't offer that reusability.

Reply:
So you don't need the boundaries then bc `Regex` semantics. But the point remains. You don't need the grouping check. Simply match and format.

Reply:
parse() factory and strict constructor. boom